File: python/xenithpay_python/signature.py
import base64
import hashlib
import hmac


class SignatureGenerator:
    def generate(
        self,
        method: str,
        uri: str,
        timestamp: str,
        body: str,
        secret_key: str,
    ) -> str:
        payload = (
            f"{method}\n"
            f"{uri}\n"
            f"{timestamp}\n"
            f"{body}"
        )

        digest = hmac.new(
            secret_key.encode("utf-8"),
            payload.encode("utf-8"),
            hashlib.sha256,
        ).digest()

        return base64.b64encode(digest).decode("utf-8")

File: php/XenithPayPhp/signature.php
<?php

declare(strict_types=1);

final class SignatureGenerator
{
    public function generate(
        string $method,
        string $uri,
        string $timestamp,
        string $body,
        string $secretKey
    ): string {
        $payload =
            $method . "\n" .
            $uri . "\n" .
            $timestamp . "\n" .
            $body;

        $hash = hash_hmac(
            'sha256',
            $payload,
            $secretKey,
            true
        );

        return base64_encode($hash);
    }
}